subsys/net/lib/nrf_cloud/coap/src/dtls.c

```diff
@@ -0,0 +1,329 @@
+		return 0;
+	}
+	if (major >= 2) {
+		mfw_ge_version_2 = true;
+		mfw_has_cid = true;
+	} else if ((major >= 1) && (minor >= 3) && (rev >= 5)) {
+		mfw_ge_version_2 = false;
+		mfw_has_cid = true;
+	}
+
+	return 0;
+}
+
+#endif /* CONFIG_MODEM_INFO */
+
+static int get_device_ip_address(uint8_t *d4_addr)
+{
+#if defined(CONFIG_MODEM_INFO)
+	int err;
+
+	err = modem_info_init();
+	if (err) {
+		return err;
+	}
+
+	err = modem_info_string_get(MODEM_INFO_IP_ADDRESS,
+				    mdm_param.network.ip_address.value_string,
+				    MODEM_INFO_MAX_RESPONSE_SIZE);
+	if (err <= 0) {
+		LOG_ERR("Could not get IP addr: %d", err);
+		return err;
+	}
+	err = inet_pton(AF_INET, mdm_param.network.ip_address.value_string, d4_addr);
+	if (err == 1) {
+		return 0;
+	}
+	return errno;
+#else
+	d4_addr[0] = 0;
+	d4_addr[1] = 0;
+	d4_addr[2] = 0;
+	d4_addr[3] = 0;
+	return 0;
+#endif
+
+}
+
+int dtls_init(int sock)
+{
+	int err;
+#if defined(RESTRICT_CIPHERS)
+	static const int ciphers[] = {
+		MBEDTLS_TLS_PSK_WITH_AES_128_CCM_8, 0
+	};
+#endif
+
+	uint8_t d4_addr[4];
+
+	/* once connected, cache the connection info */
+	dtls_cid_active = false;
+
+#if defined(CONFIG_MODEM_INFO)
+	err = get_modem_info();
+	if (err) {
+		LOG_INF("Modem firmware version not known");
+	}
+#endif
+
+	err = get_device_ip_address(d4_addr);
+	if (!err) {
+		LOG_INF("Client IP address: %u.%u.%u.%u",
+			d4_addr[0], d4_addr[1], d4_addr[2], d4_addr[3]);
+	}
+
+	LOG_INF("Setting socket options:");
+
+	LOG_INF("  hostname: %s", CONFIG_NRF_CLOUD_COAP_SERVER_HOSTNAME);
+	err = setsockopt(sock, SOL_TLS, TLS_HOSTNAME, CONFIG_NRF_CLOUD_COAP_SERVER_HOSTNAME,
+			 sizeof(CONFIG_NRF_CLOUD_COAP_SERVER_HOSTNAME));
+	if (err) {
+		LOG_ERR("Error setting hostname: %d", errno);
+		return err;
+	}
+
+	LOG_INF("  sectag: %d", sectag);
+	err = setsockopt(sock, SOL_TLS, TLS_SEC_TAG_LIST, &sectag, sizeof(sectag));
+	if (err) {
+		LOG_ERR("Error setting sectag list: %d", errno);
+		return err;
+	}
+
+#if defined(RESTRICT_CIPHERS)
+	LOG_INF("  restrict ciphers");
+	err = setsockopt(sock, SOL_TLS, TLS_CIPHERSUITE_LIST, ciphers, sizeof(ciphers));
+	if (err) {
+		LOG_ERR("Error setting cipherlist: %d", errno);
+		return err;
+	}
+#endif
+#if defined(DUMP_CIPHERLIST)
+	int len;
+	int ciphers[32];
+
+	len = sizeof(ciphers);
+	err = getsockopt(sock, SOL_TLS, TLS_CIPHERSUITE_LIST, ciphers, &len);
+	if (err) {
+		LOG_ERR("Error getting cipherlist: %d", errno);
+	} else {
+		int count = len / sizeof(int);
+
+		LOG_INF("New cipherlist:");
+		for (int i = 0; i < count; i++) {
+			LOG_INF("%d. 0x%04X = %s", i, (unsigned int)ciphers[i],
+			       IS_ENABLED(CONFIG_MBEDTLS) ?
+				mbedtls_ssl_get_ciphersuite_name(ciphers[i]) : "");
+		}
+	}
+#endif
+
+	if (mfw_has_cid) {
+		int cid_option = TLS_DTLS_CID_SUPPORTED;
+
+		LOG_INF("  Enable connection id");
+		err = setsockopt(sock, SOL_TLS, TLS_DTLS_CID, &cid_option, sizeof(cid_option));
+		if (err) {
+			LOG_ERR("Error enabling connection ID: %d", errno);
```

Comment:
ok to continue if setting TLS_DTLS_CID and/or TLS_DTLS_HANDSHAKE_TIMEO fail?
if so, the LOG_ERR could be LOG_WRN?

Reply:
If I eliminate checking modem firmware versions and just check for error returns, as pointed out by @eivindj-nordic will happen when using unsupported socket options, then yes, I could just print warnings.

Reply:
currently LOG_ERR is printed those two sock opts, but the function continues and will return success if setting the `TLS_PEER_VERIFY` sock opt succeeds.  

the intent of my comment: if it is an error, then return err, otherwise LOG_WRN seems more appropriate.

Reply:
Understood